app.go

```diff
@@ -1,248 +1,226 @@
+		log.Warn().Msg("Failed to send transaction, reason: " + sendError.Error())
+		return
+	}
+	log.Debug().Msg("Successfully signed and sent txn, id: " + result.TransactionID)
 }
 
-func (app *App) RunEventListener(parentContext context.Context, wg *sync.WaitGroup) {
-
-    go func(parentContext context.Context) {
-        defer wg.Done()
-        events := make(chan *broker.EventMessage)
-
-        listener := broker.NewEventListener(app.Broker.Url, events)
-        ctx, cancel := context.WithCancel(context.Background())
-
-        if err := listener.ListenAndServe(ctx); err != nil {
-            log.Panic().Msg(err.Error())
-        }
-
-        defer cancel()
-
-        offsetPath := app.Broker.TopicOffsetPath
-        offset := readOffset(offsetPath)
-        topicID := app.Broker.TopicID
-
-        log.Debug().Msgf("Subscribing to event type %+v with an offset of %+v", topicID, offset)
-        _, err := listener.Subscribe(topicID, offset)
-
-        if err != nil {
-            log.Error().Msg("Failed to subscribe")
-            return
-        }
-
-        for {
-           select {
-           case <-parentContext.Done():
-               log.Debug().Msg("Terminating event listener")
-               _, err := listener.Unsubscribe(topicID)
-               if err != nil {
-                   log.Warn().Msg("Failed to unsubscribe")
-               } else {
-                   log.Debug().Msg("Event listener successfully terminated")
-               }
-               return
-           case eventMessage, ok := <-events:
-               if !ok {
-                   log.Info().Msg("Failed to read events")
-                   break
-               }
-               if len(eventMessage.Events) == 0 {
-                   log.Debug().Msg("Gotta event message with no events")
-                   break
-               }
-               log.Debug().Msgf("Processing %+v events", len(eventMessage.Events))
-               for _, event := range eventMessage.Events {
-                   go app.processEvent(event)
-               }
-               offset = eventMessage.Events[len(eventMessage.Events) - 1].Offset + 1
-               writeOffset(offsetPath, offset)
-           }
-        }
-    }(parentContext)
+func (app *App) RunEventProcessor(ctx context.Context, wg *sync.WaitGroup) {
+	defer wg.Done()
+
+	for {
+		select {
+		case <-ctx.Done():
+			return
+		case eventMessage, ok := <-app.EventMessages:
+			if !ok {
+				log.Warn().Msg("Failed to read events")
+				break
+			}
+			if len(eventMessage.Events) == 0 {
+				log.Debug().Msg("Gotta event message with no events")
+				break
+			}
+			log.Debug().Msgf("Processing %+v events", len(eventMessage.Events))
+			for _, event := range eventMessage.Events {
+				go app.processEvent(event)
+			}
+			offset := eventMessage.Events[len(eventMessage.Events)-1].Offset + 1
+			if err := writeOffset(app.OffsetHandler, offset); err != nil {
+				log.Error().Msgf("Failed to write offset, reason: %s", err.Error())
```

Comment:
I think, it's fatal, if we can't write offset it potentially can bring inconsistency, so we should abort here to prevent it

Reply:
so you want to panic here?

Reply:
yup